src/ImageSharp/Configuration.cs

```diff
@@ -1,267 +1,221 @@
-                new JpegConfigurationModule(),
-                new GifConfigurationModule(),
-                new BmpConfigurationModule());
-        }
-
-        /// <summary>
-        /// Sets the max header size.
-        /// </summary>
-        private void SetMaxHeaderSize()
-        {
-            this.MaxHeaderSize = this.imageFormatDetectors.Max(x => x.HeaderSize);
-        }
-    }
-}
+﻿// Copyright (c) Six Labors and contributors.
+// Licensed under the Apache License, Version 2.0.
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using SixLabors.ImageSharp.Formats;
+using SixLabors.ImageSharp.Formats.Bmp;
+using SixLabors.ImageSharp.Formats.Gif;
+using SixLabors.ImageSharp.Formats.Jpeg;
+using SixLabors.ImageSharp.Formats.Png;
+using SixLabors.ImageSharp.IO;
+using SixLabors.ImageSharp.Memory;
+
+namespace SixLabors.ImageSharp
+{
+    /// <summary>
+    /// Provides initialization code which allows extending the library.
+    /// </summary>
+    public sealed class Configuration
+    {
+        /// <summary>
+        /// A lazily initialized configuration default instance.
+        /// </summary>
+        private static readonly Lazy<Configuration> Lazy = new Lazy<Configuration>(CreateDefaultInstance);
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="Configuration" /> class.
+        /// </summary>
+        public Configuration()
+        {
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="Configuration" /> class.
+        /// </summary>
+        /// <param name="configurationModules">A collection of configuration modules to register</param>
+        public Configuration(params IConfigurationModule[] configurationModules)
+        {
+            if (configurationModules != null)
+            {
+                foreach (IConfigurationModule p in configurationModules)
+                {
+                    p.Configure(this);
+                }
+            }
+        }
+
+        /// <summary>
+        /// Gets the default <see cref="Configuration"/> instance.
+        /// </summary>
+        public static Configuration Default { get; } = Lazy.Value;
+
+        /// <summary>
+        /// Gets the global parallel options for processing tasks in parallel.
+        /// </summary>
+        public ParallelOptions ParallelOptions { get; } = new ParallelOptions { MaxDegreeOfParallelism = Environment.ProcessorCount };
+
+        /// <summary>
+        /// Gets the currently registered <see cref="IImageFormat"/>s.
+        /// </summary>
+        public IEnumerable<IImageFormat> ImageFormats => this.FormatsManager.ImageFormats;
+
+        /// <summary>
+        /// Gets or sets the <see cref="ImageFormatsManager"/> that is currently in use.
+        /// </summary>
+        public ImageFormatsManager FormatsManager { get; set; } = new ImageFormatsManager();

```

Comment:
Can we rename `ImageFormatsManager` to `ImageFormatManager` and the property names to match please.